cmd/scheduleapi/handler/get_arrival_estimates.go

```diff
@@ -0,0 +1,186 @@
+type DynamoQuerier interface {
+	QueryPagesWithContext(ctx aws.Context, input *dynamodb.QueryInput, fn func(*dynamodb.QueryOutput, bool) bool, opts ...request.Option) error
+	QueryWithContext(ctx aws.Context, input *dynamodb.QueryInput, opts ...request.Option) (*dynamodb.QueryOutput, error)
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . GetArrivalEstimatesEndpoint
+type GetArrivalEstimatesEndpoint func(context.Context, *schedule.GetArrivalEstimatesRequest) (*schedule.GetArrivalEstimatesResponse, error)
+
+func ValidateRequest(ctx context.Context, in *schedule.GetArrivalEstimatesRequest) error {
+	var errStrings []string
+	if in == nil {
+		errStrings = append(errStrings, "request body is nil")
+	}
+	if strings.TrimSpace(in.GetDestination()) == "" {
+		errStrings = append(errStrings, "destination is nil")
+	}
+	if strings.TrimSpace(in.GetStation()) == "" {
+		errStrings = append(errStrings, "station is nil")
+	}
+	if in.GetStartDate() == nil {
+		errStrings = append(errStrings, "start date is nil")
+	}
+	if in.GetEndDate() == nil {
+		errStrings = append(errStrings, "end date is nil")
+	}
+	if in.GetStartDate().GetSeconds() > in.GetEndDate().GetSeconds() {
+		errStrings = append(errStrings, "start date must be before end date")
+	}
+	s, err := ptypes.Timestamp(in.GetStartDate())
+	if err != nil {
+		errStrings = append(errStrings, "start must be RFC3339 encoded")
+	}
+	e, err := ptypes.Timestamp(in.GetEndDate())
+	if err != nil {
+		errStrings = append(errStrings, "end must be RFC3339 encoded")
+	}
+	if e.Format("2006-01-02") != s.Format("2006-01-02") {
+		errStrings = append(errStrings, "start and end must be on the same day")
+	}
+	if len(errStrings) != 0 {
+		return errors.New(fmt.Sprintf("validation errors: %s", strings.Join(errStrings, ", ")))
+	}
+	return nil
+}
+
+func GetArrivalEstimatesRequestToDynamoQuery(in *schedule.GetArrivalEstimatesRequest, tableName string) (*dynamodb.QueryInput, error) {
+	s, err := ptypes.Timestamp(in.GetStartDate())
+	if err != nil {
+		return nil, err
+	}
+	e, err := ptypes.Timestamp(in.GetEndDate())
+	if err != nil {
+		return nil, err
+	}
+	primaryKey := fmt.Sprintf("%s_%s_%s", in.GetStation(), in.GetDestination(), s.Format("2006-01-02"))
+	keyCondition := expression.
+		Key("PrimaryKey").
+		Equal(expression.Value(primaryKey)).
+		And(expression.Key("SortKey").
+			Between(expression.Value(s.Format(time.RFC3339)), expression.Value(e.Format(time.RFC3339))))
+	expr, err := expression.NewBuilder().WithKeyCondition(keyCondition).Build()
+	if err != nil {
+		return nil, err
+	}
+
+	input := &dynamodb.QueryInput{
+		TableName:                 aws.String(tableName),
+		ExpressionAttributeNames:  expr.Names(),
+		ExpressionAttributeValues: expr.Values(),
+		KeyConditionExpression:    expr.KeyCondition(),
+	}
+	if in.GetLastEvaluatedKey() != "" {
+		lastEval := make(map[string]*dynamodb.AttributeValue)
+		decoded, err := base64.StdEncoding.DecodeString(in.GetLastEvaluatedKey())
```

Comment:
Love the base64 encoding thing! I remember reading something like this once in an article about methods for obscuring your internals when creating pagination keys. They even suggested encrypting after base64 encoding but this is more of a professional-style thing than a security concern so that seems like overkill

Reply:
Yup -- this is a good thing to do in most cases, and if this ever becomes public is definitely something we should do I think.  I'm going to keep this as `base64` here as I don't think we need to obscure it (I think `fivepoints` will always have an application layer in front of it), but nonetheless good point!!